File: python/shield/group.py
"""
Shield Group Encryption - Multi-recipient encryption.

Encrypt once for multiple recipients, each can decrypt with their own key.

Example:
    >>> from shield.group import GroupEncryption
    >>>
    >>> group = GroupEncryption()
    >>> group.add_member("alice", alice_shared_key)
    >>> group.add_member("bob", bob_shared_key)
    >>>
    >>> encrypted = group.encrypt(b"group message")
    >>>
    >>> # Each member decrypts with their key
    >>> decrypted = GroupEncryption.decrypt(encrypted, "alice", alice_shared_key)
"""

import os
import hmac
import hashlib
import struct
import secrets
import json
import base64
from typing import Dict, Optional, List


def _generate_keystream(key: bytes, nonce: bytes, length: int) -> bytes:
    """Generate keystream using SHA256."""
    keystream = b""
    for i in range((length + 31) // 32):
        keystream += hashlib.sha256(key + nonce + struct.pack("<I", i)).digest()
    return keystream[:length]


def _encrypt_block(key: bytes, data: bytes) -> bytes:
    """Encrypt a block with HMAC authentication."""
    nonce = os.urandom(16)
    keystream = _generate_keystream(key, nonce, len(data))
    ciphertext = bytes(a ^ b for a, b in zip(data, keystream))
    mac = hmac.new(key, nonce + ciphertext, hashlib.sha256).digest()[:16]
    return nonce + ciphertext + mac


def _decrypt_block(key: bytes, encrypted: bytes) -> Optional[bytes]:
    """Decrypt a block with HMAC verification."""
    if len(encrypted) < 32:
        return None
    nonce = encrypted[:16]
    ciphertext = encrypted[16:-16]
    mac = encrypted[-16:]

    expected_mac = hmac.new(key, nonce + ciphertext, hashlib.sha256).digest()[:16]
    if not hmac.compare_digest(mac, expected_mac):
        return None

    keystream = _generate_keystream(key, nonce, len(ciphertext))
    return bytes(a ^ b for a, b in zip(ciphertext, keystream))


class GroupEncryption:
    """
    Multi-recipient encryption.

    Uses a group key for message encryption, then encrypts
    the group key separately for each member.
    """

    def __init__(self, group_key: bytes = None):
        """
        Initialize group encryption.

        Args:
            group_key: 32-byte group key (generated if not provided)
        """
        self.group_key = group_key or secrets.token_bytes(32)
        self._members: Dict[str, bytes] = {}

    def add_member(self, member_id: str, shared_key: bytes) -> None:
        """
        Add a member to the group.

        Args:
            member_id: Unique member identifier
            shared_key: Pre-shared key with this member
        """
        self._members[member_id] = shared_key

    def remove_member(self, member_id: str) -> bool:
        """
        Remove a member from the group.

        Note: After removing a member, you should rotate the group key.

        Args:
            member_id: Member to remove

        Returns:
            True if member was removed
        """
        if member_id in self._members:
            del self._members[member_id]
            return True
        return False

    @property
    def members(self) -> List[str]:
        """Get list of member IDs."""
        return list(self._members.keys())

    def encrypt(self, plaintext: bytes) -> dict:
        """
        Encrypt for all group members.

        Args:
            plaintext: Message to encrypt

        Returns:
            Dictionary with ciphertext and per-member encrypted keys
        """
        # Encrypt message with group key
        ciphertext = _encrypt_block(self.group_key, plaintext)

        # Encrypt group key for each member
        encrypted_keys = {}
        for member_id, member_key in self._members.items():
            encrypted_keys[member_id] = base64.b64encode(
                _encrypt_block(member_key, self.group_key)
            ).decode('ascii')

        return {
            'version': 1,
            'ciphertext': base64.b64encode(ciphertext).decode('ascii'),
            'keys': encrypted_keys
        }

    @staticmethod
    def decrypt(encrypted: dict, member_id: str, member_key: bytes) -> Optional[bytes]:
        """
        Decrypt as a group member.

        Args:
            encrypted: Encrypted message from encrypt()
            member_id: Your member ID
            member_key: Your shared key

        Returns:
            Decrypted message, or None if decryption fails
        """
        if member_id not in encrypted.get('keys', {}):
            return None

        # Decrypt group key
        encrypted_group_key = base64.b64decode(encrypted['keys'][member_id])
        group_key = _decrypt_block(member_key, encrypted_group_key)
        if group_key is None:
            return None

        # Decrypt message
        ciphertext = base64.b64decode(encrypted['ciphertext'])
        return _decrypt_block(group_key, ciphertext)

    def rotate_key(self) -> bytes:
        """
        Rotate the group key.

        Returns:
            New group key
        """
        old_key = self.group_key
        self.group_key = secrets.token_bytes(32)
        return old_key

    def to_json(self) -> str:
        """Serialize encrypted message to JSON string."""
        # This is for the encrypt() result, not the group itself
        pass

    @staticmethod
    def from_json(json_str: str) -> dict:
        """Parse encrypted message from JSON string."""
        return json.loads(json_str)


class BroadcastEncryption:
    """
    Efficient broadcast encryption for large groups.

    Uses a key hierarchy to reduce per-message overhead.
    Members are organized into subgroups with shared subgroup keys.
    """

    def __init__(self, master_key: bytes = None, subgroup_size: int = 16):
        """
        Initialize broadcast encryption.

        Args:
            master_key: Master key for the broadcast
            subgroup_size: Members per subgroup
        """
        self.master_key = master_key or secrets.token_bytes(32)
        self.subgroup_size = subgroup_size
        self._members: Dict[str, Tuple] = {}  # member_id -> (subgroup_id, member_key)
        self._subgroup_keys: Dict[int, bytes] = {}
        self._next_subgroup = 0

    def add_member(self, member_id: str, member_key: bytes) -> int:
        """
        Add member to broadcast group.

        Args:
            member_id: Unique member ID
            member_key: Shared key with member

        Returns:
            Subgroup ID assigned
        """
        # Find subgroup with space
        subgroup_id = None
        for sg_id, sg_key in self._subgroup_keys.items():
            members_in_sg = sum(1 for m in self._members.values() if m[0] == sg_id)
            if members_in_sg < self.subgroup_size:
                subgroup_id = sg_id
                break

        if subgroup_id is None:
            subgroup_id = self._next_subgroup
            self._subgroup_keys[subgroup_id] = secrets.token_bytes(32)
            self._next_subgroup += 1

        self._members[member_id] = (subgroup_id, member_key)
        return subgroup_id

    def encrypt(self, plaintext: bytes) -> dict:
        """
        Encrypt for broadcast.

        Uses two-level encryption:
        1. Message encrypted with message key
        2. Message key encrypted with each subgroup key
        3. Subgroup keys encrypted with member keys

        Returns:
            Encrypted broadcast message
        """
        message_key = secrets.token_bytes(32)

        # Encrypt message
        ciphertext = _encrypt_block(message_key, plaintext)

        # Encrypt message key for each subgroup
        subgroup_keys_enc = {}
        for sg_id, sg_key in self._subgroup_keys.items():
            subgroup_keys_enc[str(sg_id)] = base64.b64encode(
                _encrypt_block(sg_key, message_key)
            ).decode('ascii')

        # Encrypt subgroup keys for each member
        member_keys_enc = {}
        for member_id, (sg_id, member_key) in self._members.items():
            sg_key = self._subgroup_keys[sg_id]
            member_keys_enc[member_id] = {
                'sg': sg_id,
                'key': base64.b64encode(_encrypt_block(member_key, sg_key)).decode('ascii')
            }

        return {
            'version': 1,
            'ciphertext': base64.b64encode(ciphertext).decode('ascii'),
            'subgroups': subgroup_keys_enc,
            'members': member_keys_enc
        }

    @staticmethod
    def decrypt(encrypted: dict, member_id: str, member_key: bytes) -> Optional[bytes]:
        """
        Decrypt broadcast as member.

        Args:
            encrypted: Encrypted broadcast
            member_id: Your member ID
            member_key: Your shared key

        Returns:
            Decrypted message
        """
        if member_id not in encrypted.get('members', {}):
            return None

        member_data = encrypted['members'][member_id]
        sg_id = member_data['sg']

        # Decrypt subgroup key
        sg_key_enc = base64.b64decode(member_data['key'])
        sg_key = _decrypt_block(member_key, sg_key_enc)
        if sg_key is None:
            return None

        # Decrypt message key
        msg_key_enc = base64.b64decode(encrypted['subgroups'][str(sg_id)])
        msg_key = _decrypt_block(sg_key, msg_key_enc)
        if msg_key is None:
            return None

        # Decrypt message
        ciphertext = base64.b64decode(encrypted['ciphertext'])
        return _decrypt_block(msg_key, ciphertext)

File: go/shield/group.go
package shield

import (
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"encoding/hex"
	"errors"
)

var (
	// ErrMemberNotFound indicates member not found in group.
	ErrMemberNotFound = errors.New("shield: member not found")
	// ErrNoRecipients indicates no recipients for broadcast.
	ErrNoRecipients = errors.New("shield: no recipients")
)

// GroupEncryption provides multi-party group encryption.
type GroupEncryption struct {
	groupKey  [KeySize]byte
	memberKey [KeySize]byte
	memberID  string
	members   map[string][]byte // memberID -> encryptedGroupKey
}

// NewGroupEncryption creates a new group with this member as admin.
func NewGroupEncryption(memberID string, memberKey []byte) (*GroupEncryption, error) {
	if len(memberKey) != KeySize {
		return nil, ErrInvalidKeySize
	}

	ge := &GroupEncryption{
		memberID: memberID,
		members:  make(map[string][]byte),
	}
	copy(ge.memberKey[:], memberKey)

	// Generate random group key
	if _, err := rand.Read(ge.groupKey[:]); err != nil {
		return nil, err
	}

	// Encrypt group key for self
	encryptedKey, err := EncryptWithKey(memberKey, ge.groupKey[:])
	if err != nil {
		return nil, err
	}
	ge.members[memberID] = encryptedKey

	return ge, nil
}

// JoinGroup joins an existing group.
func JoinGroup(memberID string, memberKey, encryptedGroupKey []byte) (*GroupEncryption, error) {
	if len(memberKey) != KeySize {
		return nil, ErrInvalidKeySize
	}

	ge := &GroupEncryption{
		memberID: memberID,
		members:  make(map[string][]byte),
	}
	copy(ge.memberKey[:], memberKey)

	// Decrypt group key
	groupKey, err := DecryptWithKey(memberKey, encryptedGroupKey, nil)
	if err != nil {
		return nil, err
	}
	copy(ge.groupKey[:], groupKey)
	ge.members[memberID] = encryptedGroupKey

	return ge, nil
}

// AddMember adds a member to the group.
func (ge *GroupEncryption) AddMember(memberID string, memberKey []byte) ([]byte, error) {
	if len(memberKey) != KeySize {
		return nil, ErrInvalidKeySize
	}

	// Encrypt group key for new member
	encryptedKey, err := EncryptWithKey(memberKey, ge.groupKey[:])
	if err != nil {
		return nil, err
	}
	ge.members[memberID] = encryptedKey

	return encryptedKey, nil
}

// RemoveMember removes a member (does NOT revoke their key).
func (ge *GroupEncryption) RemoveMember(memberID string) error {
	if _, ok := ge.members[memberID]; !ok {
		return ErrMemberNotFound
	}
	delete(ge.members, memberID)
	return nil
}

// Encrypt encrypts a message for the group.
func (ge *GroupEncryption) Encrypt(plaintext []byte) ([]byte, error) {
	return EncryptWithKey(ge.groupKey[:], plaintext)
}

// Decrypt decrypts a group message.
func (ge *GroupEncryption) Decrypt(ciphertext []byte) ([]byte, error) {
	return DecryptWithKey(ge.groupKey[:], ciphertext, nil)
}

// GroupKey returns the group key (for backup/admin).
func (ge *GroupEncryption) GroupKey() []byte {
	return ge.groupKey[:]
}

// Members returns list of member IDs.
func (ge *GroupEncryption) Members() []string {
	members := make([]string, 0, len(ge.members))
	for id := range ge.members {
		members = append(members, id)
	}
	return members
}

// Fingerprint returns the group key fingerprint.
func (ge *GroupEncryption) Fingerprint() string {
	h := sha256.Sum256(ge.groupKey[:])
	return hex.EncodeToString(h[:8])
}

// BroadcastEncryption provides one-to-many encryption.
type BroadcastEncryption struct {
	senderKey  [KeySize]byte
	recipients map[string][]byte // recipientID -> recipientKey
}

// NewBroadcastEncryption creates a new broadcast sender.
func NewBroadcastEncryption(senderKey []byte) (*BroadcastEncryption, error) {
	if len(senderKey) != KeySize {
		return nil, ErrInvalidKeySize
	}

	be := &BroadcastEncryption{
		recipients: make(map[string][]byte),
	}
	copy(be.senderKey[:], senderKey)

	return be, nil
}

// AddRecipient adds a recipient.
func (be *BroadcastEncryption) AddRecipient(recipientID string, recipientKey []byte) error {
	if len(recipientKey) != KeySize {
		return ErrInvalidKeySize
	}
	keyCopy := make([]byte, KeySize)
	copy(keyCopy, recipientKey)
	be.recipients[recipientID] = keyCopy
	return nil
}

// RemoveRecipient removes a recipient.
func (be *BroadcastEncryption) RemoveRecipient(recipientID string) {
	delete(be.recipients, recipientID)
}

// BroadcastMessage represents a broadcast encrypted message.
type BroadcastMessage struct {
	EncryptedMessage []byte
	RecipientKeys    map[string][]byte // recipientID -> encrypted session key
	SenderProof      []byte
}

// Broadcast encrypts a message for all recipients.
func (be *BroadcastEncryption) Broadcast(plaintext []byte) (*BroadcastMessage, error) {
	if len(be.recipients) == 0 {
		return nil, ErrNoRecipients
	}

	// Generate random session key
	sessionKey := make([]byte, KeySize)
	if _, err := rand.Read(sessionKey); err != nil {
		return nil, err
	}

	// Encrypt message with session key
	encryptedMsg, err := EncryptWithKey(sessionKey, plaintext)
	if err != nil {
		return nil, err
	}

	// Encrypt session key for each recipient
	recipientKeys := make(map[string][]byte)
	for id, key := range be.recipients {
		encKey, err := EncryptWithKey(key, sessionKey)
		if err != nil {
			return nil, err
		}
		recipientKeys[id] = encKey
	}

	// Create sender proof
	mac := hmac.New(sha256.New, be.senderKey[:])
	mac.Write(encryptedMsg)
	proof := mac.Sum(nil)

	return &BroadcastMessage{
		EncryptedMessage: encryptedMsg,
		RecipientKeys:    recipientKeys,
		SenderProof:      proof,
	}, nil
}

// ReceiveBroadcast decrypts a broadcast message.
func ReceiveBroadcast(msg *BroadcastMessage, recipientID string, recipientKey []byte) ([]byte, error) {
	encryptedSessionKey, ok := msg.RecipientKeys[recipientID]
	if !ok {
		return nil, ErrMemberNotFound
	}

	// Decrypt session key
	sessionKey, err := DecryptWithKey(recipientKey, encryptedSessionKey, nil)
	if err != nil {
		return nil, err
	}

	// Decrypt message
	return DecryptWithKey(sessionKey, msg.EncryptedMessage, nil)
}

// VerifySender verifies the sender proof.
func VerifySender(msg *BroadcastMessage, senderKey []byte) bool {
	mac := hmac.New(sha256.New, senderKey)
	mac.Write(msg.EncryptedMessage)
	expected := mac.Sum(nil)
	return hmac.Equal(msg.SenderProof, expected)
}

// Recipients returns list of recipient IDs.
func (be *BroadcastEncryption) Recipients() []string {
	recipients := make([]string, 0, len(be.recipients))
	for id := range be.recipients {
		recipients = append(recipients, id)
	}
	return recipients
}
